Paper at: www.nature.com/articles/s41588-018-0167-z

Using hundreds of thousands of common variants from population sequencing of six non-human primate species, PrimateAI is a trained deep neural network to classify variants' pathogenicity.  It predicts impact purely from sequence and predicts secondary structure information and solvent accessibility information.

From the paper:

"Prediction scores for all ~70 million human missense variants on the hg19/GRCh37 genome build with the human+primate deep learning network (PrimateAI) are publicly hosted (https://basespace.illumina.com/s/cPgCSmecvhb4). For practical application of PrimateAI scores, we recommend a threshold of >0.8 for likely pathogenic classification, <0.6 for likely benign, and 0.6-0.8 as intermediate in genes with dominant modes of inheritance, on the basis of the enrichment of de novo variants in cases as compared to controls (Fig. 3d), and a threshold of >0.7 for likely pathogenic and <0.5 for likely benign in genes with recessive modes of inheritance."
